sql/schemachanger: preserve some column attributes during data type alter

spilchen opened the pull request:
Previously, altering a data type on a column with the hidden attribute would succeed, but the column would lose the hidden attribute. Additionally, if a column had a DEFAULT or ON UPDATE expression, the ALTER would fail.

To address this:

- ON UPDATE and DEFAULT expressions are now deferred until we swap the computed property of the column with the shadow column. These expressions can't be set when the shadow column is created because it starts as a computed column, which can't have these expressions.
- The hidden attribute is now copied over when creating the shadow column, preserving it properly.

Epic: CRDB-25314
Closes: #71571
Release note (bug fix): When altering the data type of columns with the hidden attribute, we now preserve that in the alter column. Also, we now allow type alterations for columns with ON UPDATE/DEFAULT expressions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126033)
<!-- Reviewable:end -->

fqazi reviewed (approved):
Great work, just one minor comment


<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-:-O0GSjTpD3PvtEvsxGnj:b-ejxdjx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom and @spilchen)

___
*[`pkg/sql/catalog/tabledesc/structured.go` line 1685 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GRwrsDhJbjxg_MfHY:-O0GRwrsDhJbjxg_MfHZ:b-bxfsfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/catalog/tabledesc/structured.go#L1685)):*
> ```Go
> 	// for non-computed columns, so it needs to be swapped at the same time as the
> 	// computed expression.
> 	newCol.ColumnDesc().OnUpdateExpr = oldCol.ColumnDesc().OnUpdateExpr
> ```

I think we already validate that these types can be cast?
___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0GSdGjAhof0RPj_JJL:b-2ns9li) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
> ```
> 
> statement ok
> DROP TABLE t_with_default;
> ```

Should we do a SHOW CREATE?


<!-- Sent from Reviewable.io -->

annrpom reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/catalog/tabledesc/structured.go` line 1686 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GT7U-ClRl69bfoL8i:-O0GT7U-ClRl69bfoL8j:b-vu010f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/catalog/tabledesc/structured.go#L1686)):*
> ```Go
> 	// computed expression.
> 	newCol.ColumnDesc().OnUpdateExpr = oldCol.ColumnDesc().OnUpdateExpr
> 	oldCol.ColumnDesc().OnUpdateExpr = nil
> ```

I see the following (and might be confused):
1. that we set oldCol to be a computed column up above
2. that we just unset oldCol's onupdate/default exprs in this PR
3. computed columns can't have onupdate/default exprs

So does that mean before this PR, we allowed for on update/default exprs to be set on computed columns (if so, could we add a regression test and an update to the commit message/release note)?
___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0GUGmODRXHo_M88gTS:b-m6ywg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Should we do a SHOW CREATE?
</blockquote></details>

+1 the crowd screams "show create please"


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: preserve some column attributes during data type alter

Previously, altering a data type on a column with the hidden attribute
would succeed, but the column would lose the hidden attribute.
Additionally, if a column had a DEFAULT or ON UPDATE expression, the
ALTER would fail.

To address this:

- ON UPDATE and DEFAULT expressions are now deferred until we swap the
  computed property of the column with the shadow column. These
  expressions can't be set when the shadow column is created because it
  starts as a computed column, which can't have these expressions.
- The hidden attribute is now copied over when creating the shadow
  column, preserving it properly.

Epic: CRDB-25314
Closes: #71571
Release note (bug fix): When altering the data type of columns with
the hidden attribute, we now preserve that in the alter column. Also, we
now allow type alterations for columns with ON UPDATE/DEFAULT expressions.

spilchen reviewed (commented):
tftr!

I responded to your comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-:-O0K2V1-0KV2E09tMzrx:b-nkposd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/catalog/tabledesc/structured.go` line 1685 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GRwrsDhJbjxg_MfHY:-O0JmnJe7cXZBkqD95KF:b-e32ivd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/catalog/tabledesc/structured.go#L1685)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think we already validate that these types can be cast?
</blockquote></details>

Yes, we validate the cast in `AlterColumnType`. I will clarify this in the comment.
___
*[`pkg/sql/catalog/tabledesc/structured.go` line 1686 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GT7U-ClRl69bfoL8i:-O0JoKyB5cCbhtnzczW_:b3pkj5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/catalog/tabledesc/structured.go#L1686)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

I see the following (and might be confused):
1. that we set oldCol to be a computed column up above
2. that we just unset oldCol's onupdate/default exprs in this PR
3. computed columns can't have onupdate/default exprs

So does that mean before this PR, we allowed for on update/default exprs to be set on computed columns (if so, could we add a regression test and an update to the commit message/release note)?
</blockquote></details>

Your reasoning is correct. But my commit message is incorrect as I don't correctly explain the current behaviour. That's probably were the confusion lies.

The current behaviour, which is different than what the GH issue says, is that the alter would fail if one of these expressions are set. For example:
```
SET enable_experimental_alter_column_type_general = true;
create table t(a int primary key, b int on update 3);
alter table t alter column b set data type int4;
ERROR: relation "t" (112): computed column "b_1_1" cannot also have an ON UPDATE expression
SQLSTATE: 42P16
```

Now, I am simply allowing the alter to proceed by correctly handling the expressions. I'll update the commit message.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0Jlp3Y4GEvw7jcHp8v:b-fhtzo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

+ 1 the crowd screams "show create please"
</blockquote></details>

Yes, I will add the SHOW CREATE. When I did this, I noticed that the assignment of columns to families seemed arbitrary. Sometimes, a single family would be created with all the columns, while other times, the columns would be split across multiple families. Is this behaviour expected? I am considering opening an issue for this. To work around it in this pull request, I made the families explicit in the `CREATE TABLE` statements I added.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0K5p7o2XsnkefCqwDK:b378h7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
> When I did this, I noticed that the assignment of columns to families seemed arbitrary

Hm -- this is also unexpected behavior from my end. My understanding is that by default, all columns should be stored as a single column family. I also can't find feature/PR that mentions an automatic regrouping of column families


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0KI7kdA8pvG7tdckPf:b89xsrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

> When I did this, I noticed that the assignment of columns to families seemed arbitrary

Hm -- this is also unexpected behavior from my end. My understanding is that by default, all columns should be stored as a single column family. I also can't find feature/PR that mentions an automatic regrouping of column families
</blockquote></details>

I opened issue 126265 for the issue with column families.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0KJbZY2odF5R3fruC7:bt54cap) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I opened issue 126265 for the issue with column families.
</blockquote></details>

We should be maintaining the same one in this case. Is he problem visible before the type change too?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0KKlz_AiPM2M8Y2CxS:b-jspe2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
> When I did this, I noticed that the assignment of columns to families seemed arbitrary. Sometimes, a single family would be created with all the columns, while other times, the columns would be split across multiple families. Is this behaviour expected?

yes, there is a component of the logic test framework that adds column families to CREATE statements in a randomized fashion. in most cases, column families should not affect functional behavior, so this provides test coverage. in the case of testing SHOW CREATE TABLE, we should opt-out of that behavior by explicitly adding a column family definition to the CREATE statement in the test.

see: https://github.com/cockroachdb/cockroach/blob/c50ccf674d2639adf757f0b35a91515fc3d75ba5/pkg/sql/logictest/logic.go#L3419


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126033#-O0GSdGjAhof0RPj_JJK:-O0KQ4_L9NGybvO0rREr:b-fhmxfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec3d4216bdfc2d9cce9a0037cc6dff457f719f3/pkg/sql/logictest/testdata/logic_test/alter_column_type#L798)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> When I did this, I noticed that the assignment of columns to families seemed arbitrary. Sometimes, a single family would be created with all the columns, while other times, the columns would be split across multiple families. Is this behaviour expected?

yes, there is a component of the logic test framework that adds column families to CREATE statements in a randomized fashion. in most cases, column families should not affect functional behavior, so this provides test coverage. in the case of testing SHOW CREATE TABLE, we should opt-out of that behavior by explicitly adding a column family definition to the CREATE statement in the test.

see: https://github.com/cockroachdb/cockroach/blob/c50ccf674d2639adf757f0b35a91515fc3d75ba5/pkg/sql/logictest/logic.go#L3419
</blockquote></details>

Okay, that explains it then. Thanks Rafi. I'll return the issue I opened.


<!-- Sent from Reviewable.io -->

spilchen commented:
tftr!

bors r+
